### ISSUE TITLE
Push Notifications

### ISSUE THREAD

**Issue by scotthooker** [signalpoint/DrupalGap#96]
More of a discussion than an issue but I've successfully implemented the phonegap push notification plugin along with the drupal push notification module and have this working in my drupalgap app.

Will look to see how I can make the code generic :) it's not very specific at the moment and I think will be a nice addition to the framework.

**Comment by signalpoint** [comment]
Thanks, sounds like an interesting project. I'll be curious to see what type of code was required to make it happen. Then we can figure out a way to turn it into a contrib module for DG.

**Comment by scotthooker** [comment]
It's actually not as hard as I thought and I have it all working as I said.

What will make it workable in the DG world is having somewhere to hook in the the call to store the device token. 

My problem was that it makes little sense for DG to have notifications sent to user 0. In almost all scenarios I'm using DG for its for people to login to drupal on there phone. So storing the device token on deviceready wont work as you can't rely on the user being logged in.

Thus I propose we introduce a hook/event into the login code so modules can react once the user has logged in successfully.

From that I can then call the right service to store the phones device token in the database against a user ID enabling push notifications to that device/user. 

Similarly on log out I can kill the token from the drupal DB so resources arn't wasted pushing notifications to people who won't receive them.

**Comment by signalpoint** [comment]
@scotthooker - this should currently be possible with hook_services_success().

Check out the first DG contributed module for example usage:

https://github.com/signalpoint/force_authentication/blob/7.x-1.x/force_authentication.js

Please be aware that this hook will be evolving very soon. Once this branch is merged back in, the way we implement the hook_services_success() function will be slightly different.

https://github.com/signalpoint/DrupalGap/tree/7.x-1.x-alpha-service-hooks

**Comment by scotthooker** [comment]
So I can see if the service call is login and then if it is and we are logged in I can fire my code to register the token?

**Comment by signalpoint** [comment]
@scotthooker - correct.

**Comment by signalpoint** [comment]
@scotthooker - Hey Scott, how did the Push Notifications work out for you?

**Comment by scotthooker** [comment]
Yeh all working with phonegap and phonegap build with drupalgap. I'm away this week but will share my got repo when I return and you can see it was quite simple.

**Comment by chris-jamieson** [comment]
Hi @scotthooker - sorry to post here but I'm really interested in taking a look at using Push Notifications in DG, any chance of seeing your code?

**Comment by scotthooker** [comment]
Hey man,

No worries I've been meaning to make it into generic module and release it but time constraints and all that! Plus I always write better code when people are going to see it ;) 

Do you have Skype? I can share with you some of the bits and probably will cause me to think about a few things I could do better!

**Comment by chris-jamieson** [comment]
Yeah I have Skype (jamiesoncj is username) - would be great to check this out!

**Comment by signalpoint** [comment]
@scotthooker and @jamiesoncj - please let me know how this goes, I'd be glad to help turn this into a contrib module for DG, thanks!

**Comment by scotthooker** [comment]
We went through it on Skype which caused me to think about a few things! 
I'll happily abstract the code and write it up so we can look into how to make it a contrib module. I'll get to work over the weekend :)

**Comment by signalpoint** [comment]
Resolved, thanks to @scotthooker.

Check out the Push Notifications module for DrupalGap: http://drupalgap.org/project/push_notifications

**Comment by scotthooker** [comment]
Thanks. 

Should I be using hook_install rather than hook_deviceready in my push module similar to how you do it in the force module?

**Comment by signalpoint** [comment]
When hook_install() is fired, the System Connect hasn't fired yet, so Drupal.user will always appear anonymous. I'm using hook_services_request_pre_postprocess_alter() in Force Authentication, but I think you can use hook_services_postprocess().

http://www.drupalgap.org/node/202

Definitely don't use hook_deviceready() or hook_install() though, Drupal.user is always anonymous for both of those. I've updated the API documentation to mention this.

**Comment by daode** [comment]
Hello, is there un example of this module, where can i find documents or tutoriels about the realisation of this feature. Thanks.

**Comment by alexschedrov** [comment]
The hardest, for push notifications, it's setup sending notifications from Drupal site. Certificates for Apple Notifications Server and Google Cloud Messaging... That would be nice to have documentation about this, namely for DrupalGap.

**Comment by scotthooker** [comment]
Hi all,

I created this module but haven't had time to refactor it and write come docukentation. Happy to work with someone to complete this.

**Comment by scotthooker** [comment]
Hi all,

I created this module but haven't had time to refactor it and write come docukentation. Happy to work with someone to complete this.

**Comment by daode** [comment]
Thanks for the big work to create this module ;) ,
If you have a simple example of using or screenshots, that would be a big plus.
if you have time, a small procedure with points would be welcome.
